core/corescan/context.go

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright 2021-2024 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package corescan
+
+import (
+	"fmt"
+	"github.com/JetBrains/qodana-cli/v2024/platform/msg"
+	"github.com/JetBrains/qodana-cli/v2024/platform/product"
+	"github.com/JetBrains/qodana-cli/v2024/platform/qdyaml"
+	"math"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+const (
+	RunScenarioDefault      = "default"
+	RunScenarioFullHistory  = "full-history"
+	RunScenarioLocalChanges = "local-changes"
+	RunScenarioScoped       = "scope"
+)
+
+type RunScenario = string
+
+// Context
+//
+// !!!KEEP IT IMMUTABLE!!!
+// !!!KEEP IT IMMUTABLE!!!
+// !!!KEEP IT IMMUTABLE!!!
+//
+// If one has the instance of Context, then it means that it was initialized, and it is in valid state
+// all mutations should be defined in context_changes.go with names clearly demonstrating the usecase and business logic.
+// example: scoped script launches two stages of default analysis
+//
+// # In the future, you can consider making it an interface
+//
+// Yes, there is a lot of boilerplate, and it's ok, it's much better than having one object written and initialized across your program
+// if I have an object it means it must be in valid state, otherwise it's impossible to reason about the code
+//
+// Q: - Why do we prohibit mutations in Context if we pass it by value already?
+// A: - Because we want to prohibit any unexpected changes to Context at all, imagine at some place the projectDir will
+//
+//	suddenly change, and pass it further? It's not clear why it was changed? Was it actually initialized at this place?
+//	but it must be initialized already, so we limit all changes, and keep them in context_changes.go
+//
+// !!!KEEP IT IMMUTABLE!!!
+// !!!KEEP IT IMMUTABLE!!!
+// !!!KEEP IT IMMUTABLE!!!
+type Context struct {
```

Comment:
Instead of single QodanaOptions, now this is a key struct with data for analysis, it's immutable. Constructed only when all data present for it is available.
For thirdparty linters analysis there is a different context